### ISSUE TITLE
[clang] -Wcast-qual does not trigger on functional casts

### ISSUE THREAD

**Issue by killerbee13** [llvm/llvm-project#62083]
`-Wcast-qual` does not trigger on the following code in Clang, but does in GCC.

```cpp
const auto i = 42;
using T = int*;
auto p = T(&i);
```

The expected behavior is that a functional cast should trigger the warning the same as the equivalent C cast because the meaning is the same, and nothing about the functional cast makes it easier to recognize that a const_cast is occurring (in fact I would argue it is the opposite) so the same rationale should apply.

**Comment by shafik** [comment]
So in `Sema::BuildCStyleCastExpr` it calls:

```cpp
// -Wcast-qual
  DiagnoseCastQual(Op.Self, Op.SrcExpr, Op.DestType);
```

It looks like all we need to do is add the same call in `ema::BuildCXXFunctionalCastExpr`

@AaronBallman am I missing any reason why we might not want to do that?

**Comment by AaronBallman** [comment]
> @AaronBallman am I missing any reason why we might not want to do that?

I don't think so; GCC diagnoses this under `-Wcast-qual` and has since... forever-ish. This may very well be an oversight (unless it turns out there's a high rate of false positives for some reason I can't think of).
